flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/nodes/resource/parallelism/FinalParallelismSetter.java

```diff
@@ -0,0 +1,135 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.plan.nodes.resource.parallelism;
+
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.transformations.StreamTransformation;
+import org.apache.flink.table.plan.nodes.exec.ExecNode;
+import org.apache.flink.table.plan.nodes.physical.batch.BatchExecBoundedStreamScan;
+import org.apache.flink.table.plan.nodes.physical.batch.BatchExecTableSourceScan;
+import org.apache.flink.table.plan.nodes.physical.stream.StreamExecDataStreamScan;
+import org.apache.flink.table.plan.nodes.physical.stream.StreamExecTableSourceScan;
+
+import org.apache.calcite.rel.RelDistribution;
+import org.apache.calcite.rel.core.Exchange;
+import org.apache.calcite.rel.core.Values;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Set final parallelism if needed at the beginning time, if parallelism of a node is set to be final,
+ * it will not be changed by other parallelism calculator.
+ */
+public class FinalParallelismSetter {
+
+	private final StreamExecutionEnvironment env;
+	private Set<ExecNode<?, ?>> calculatedNodeSet = new HashSet<>();
+	private Map<ExecNode<?, ?>, Integer> finalParallelismNodeMap = new HashMap<>();
+
+	private FinalParallelismSetter(StreamExecutionEnvironment env) {
+		this.env = env;
+	}
+
+	/**
+	 * Finding nodes that need to set final parallelism.
+	 */
+	public static Map<ExecNode<?, ?>, Integer> calculate(StreamExecutionEnvironment env, List<ExecNode<?, ?>> sinkNodes) {
+		FinalParallelismSetter setter = new FinalParallelismSetter(env);
+		sinkNodes.forEach(setter::calculate);
+		return setter.finalParallelismNodeMap;
+	}
+
+	private void calculate(ExecNode<?, ?> execNode) {
+		if (!calculatedNodeSet.add(execNode)) {
+			return;
+		}
+		if (execNode instanceof BatchExecTableSourceScan) {
+			calculateTableSource((BatchExecTableSourceScan) execNode);
+		} else if (execNode instanceof StreamExecTableSourceScan) {
+			calculateTableSource((StreamExecTableSourceScan) execNode);
+		} else if (execNode instanceof BatchExecBoundedStreamScan) {
+			calculateBoundedStreamScan((BatchExecBoundedStreamScan) execNode);
+		} else if (execNode instanceof StreamExecDataStreamScan) {
+			calculateDataStreamScan((StreamExecDataStreamScan) execNode);
+		} else if (execNode instanceof Values) {
+			calculateValues(execNode);
+		} else {
+			calculateIfSingleton(execNode);
+		}
+	}
+
+	private void calculateTableSource(BatchExecTableSourceScan tableSourceScan) {
+		StreamTransformation transformation = tableSourceScan.getSourceTransformation(env);
+		if (transformation.getMaxParallelism() > 0) {
+			tableSourceScan.getResource().setMaxParallelism(transformation.getMaxParallelism());
```

Comment:
It seems that the max parallelism of source transformation is not used to avoid set an invalid parallelism like what we do for sink?

Reply:
Calculating parallelism of shuffleStage later in ShuffleStageParallelismCalculator considers the max parallelism of exec nodes.